README.md
![](https://i.postimg.cc/D0w0d7kB/mmm-ratp.png)

# MMM-RATP

MMM-RATP is a MagicMirror² module that allows you to get waiting times for specific stations, alongside traffic information for the lines you want. This project is not officially supported nor endorsed by the RATP company, I made it for my own needs and thought I would share it.

It uses the gorgeous [RATP API](https://api-ratp.pierre-grimaud.fr/v4/) from Pierre Grimaud internally, you'll need to check it out to find the identifiers of your stations.

**Features**

* [x] Get as much as the next 4 **passing times for whatever station and line you want** (compatible with metros, RERs, tramways, and buses)
* [x] Get **traffic information** about your line. Is Metro 13 having slowdowns again? You will know!
* [x] A **sleek design** right?
* [x] Two bundled themes: a black/white one for mirrors, and a colored one for home dashboards.
* [x] Support for **multiple instances**. You want to have the next passes on the left and traffic information on the right? You can!
* [x] **Server-side data parsing**. Makes refreshes blazingly fast!
* [x] Dedicated update intervals. I don't think you need to refresh traffic information every minute!
* [x] **Predictive estimations!** Did an API call fail for some reasons? The module will try to estimate the next pass based on previous data
* [x] Debugging you can enable directly in the config

## Installation

There is only one dependency for this module which is [phin](https://www.npmjs.com/package/phin), an "ultra-lightweight Node.js HTTP client".

```sh
$ cd ~/MagicMirror/modules
$ git clone https://gitlab.com/closingin/MMM-RATP
$ cd MMM-RATP
$ npm i
```

## Usage

```
{
    module: 'MMM-RATP',
    position: 'bottom_right',
    config: {
        theme: 'dashboard',
        timetables: {
            config: [
                { type: 'metro', line: '14', station: 'pyramides', direction: 'A' },
            ]
        },
        traffic: {
            config: [
                { type: 'metro', line: '14' }
            ]
        }
    }
}
```

## Configuration options

### General options

| Property              | Type      | Description |
| --------------------- | --------- | ----------- |
| `theme`               | `String`  | The theme to use on the module (see screenshot above).<br><br>**Possible values:** `mirror`, `dashboard`<br>**Default value:** `mirror` |
| `debug`               | `Boolean` | Enables a debug mode that makes the module log more.<br><br>**Possible values:** `true`, `false`<br>**Default value:** `false` |
| `showUpdateAnimation` | `Boolean` | Displays a small spinner at the end of the section title when refreshing data.<br><br>**Possible values:** `true`, `false`<br>**Default value:** `true` |
| `timetables`          | `Object`  | The configuration options for your timetables, see [section](https://gitlab.com/closingin/mmm-ratp#timetables-configuration) below. |
| `traffic`             | `Object`  | The configuration options for traffic information, see [section](https://gitlab.com/closingin/mmm-ratp#traffic-information-configuration) below. |

#### Timetables configuration

The `timetables` option allows you to define the configuration of the timetables you want to display on your screen, it's an object that can contain the options below. Only `config` is mandatory.

| Property           | Type             | Description |
| ------------------ | ---------------- | ----------- |
| `title`            | `String`         | The title to display above the timetables section.<br><br>**Default value:** `'Prochains passages'` |
| `updateInterval`   | `Number`         | The time to wait between each timetables refresh, in milliseconds.<br><br>**Default value:** `1 * 60 * 1000` (1 minute) |
| `nextPassesAmount` | `Number`         | The maximum amount of upcoming passes to display (you might actually see less than the amount you ask, it depends on the data that the API provides for your line).<br><br>**Possible values:** `1`, `2`, `3`, or `4`<br>**Default value:** `2` |
| `config`           | `Array<Object>`  | The definition of the timetables you want. Look at the example usage above to understand how it works. A reference will be available below.<br><br>**Required properties for each object:** `type`, `line`, `station`, `direction` |

#### Traffic information configuration

The `traffic` option allows you to define the configuration of the traffic information you want to display on your screen, it's an object that can contain the options below. Only `config` is mandatory.

| Property         | Type            | Description |
| ---------------- | --------------- | ----------- |
| `title`          | `String`        | The title to display above the traffic information section.<br><br>**Default value:** `'Infos trafic'` |
| `updateInterval` | `Number`        | The time to wait between each traffic information refresh, in milliseconds.<br><br>**Default value:** `10 * 60 * 1000` (10 minutes) |
| `hideWhenNormal` | `Boolean`       | Whether to hide traffic information for lines that are running normally. A dedicated message will be displayed if all of your lines are running normally.<br><br>**Possible values:** `true` or `false`<br>**Default value:** `false` |
| `config`         | `Array<Object>` | The definition of the timetables you want. Look at the example usage above to understand how it works. A reference will be available below.<br><br>**Required properties for each object:** `type`, `line` |

### Reference for the timetables and traffic information `config` object

| Property    | Type     | Description |
| ----------- | -------- | ----------- |
| `type`      | `String` | The line type.<br><br>**Possible values:** `'bus'`, `'metro'`, `'rer'`, `'tramway'`<br><br>Technically the API also supports noctiliens, but this module is not yet ready for it. Soon. |
| `line`      | `String` | The line identifier.<br><br>**Examples:** `'2'` (metro), `'3b'`(tramway), `'B'` (rer) |
| `station`   | `String` | The station slug, you have to get it directly from [Pierre's API](https://api-ratp.pierre-grimaud.fr/v4). Here's an example URL that will give you the list of stations for Metro 9 : [https://api-ratp.pierre-grimaud.fr/v4/stations/metros/9](https://api-ratp.pierre-grimaud.fr/v4/stations/metros/9)<br><br>**Examples:** `'place+de+clichy'`, `'porte+de+clichy+++tribunal+de+paris'` |
| `direction` | `String` | For which direction you want to get the next passes.<br><br>**Possible values:** `'A'`, `'R'` (it stands for Aller / Retour) |

## Troubleshooting

I created this module using `nodejs@14.5.0` (MagicMirror² requires `nodejs@10`), so don't hesitate to open an issue if you have problems setting up this module.


CHANGELOG.md
# 1.1.1

- **bugfix**: display incoming train timer correctly

# 1.1.0

- **feature**: add support for buses
- **bugfix**: improve the detection of vehicles arriving at the station

# 1.0.2

- **bugfix**: handle waiting times formatted as HH:mm in some cases

# 1.0.1

- **feature**: add support for nvm and webstorm
- **bugfix**: stop displaying sections that are not configured (#1)
- **refactor**: ease theme development by moving line color declarations
- **style**: use snake case on js filenames
- **docs**: add missing jsdoc and improve existing docs

# 1.0.0 - First release!

- **feature**: spinner animation when updating data
- **feature**: add support for multiple instances
- **feature**: add a debug mode
- **bugfix**: handle lines that are stopped (e.g. during the night)
- **bugfix**: make the timer text transparent when train arriving
- **docs**: globally improve code documentation

# 0.1.0

- **feature**: get timetables for your stations
- **feature**: get traffic information
- **feature**: dedicated update intervals
- **feature**: black/white theme
- **feature**: colored theme
- **feature**: custom amount of passing times to display
- **feature**: hide traffic information for lines in normal run
- **feature**: predictive estimations


js/Utils.js
/*
 * This file is part of MMM-RATP (https://gitlab.com/closingin/mmm-ratp)
 * Copyright (C) 2020 Rémi Weislinger
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * debug - Displays a given text in the console if debug mode is enabled
 *
 * @param {Boolean} isDebugEnabled   Whether debug mode is enabled or not
 * @param {String}  moduleIdentifier The module identifier
 * @param {...*}    data             A variable amount of data to display in the message
 *
 * @returns {void} This function doesn't return anything
 */
function debug (isDebugEnabled, moduleIdentifier, ...data) {
  if (!isDebugEnabled) {
    return;
  }

  console.debug(`[MMM-RATP][DEBUG][${moduleIdentifier}]`, ...data);
}

/**
 * notificationTypeToLower - Retrieve a notification type and transform it to lower case
 *
 * @param {String} notification The notification
 *
 * @returns {String} The notification type, in lower case
 */
function notificationTypeToLower (notification) {
  return notification.split('_')[1].toLowerCase();
}

/**
 * notificationTypeToSentence - Retrieve a notification type and transform it to sentence case
 *
 * @param {String} notification The notification
 *
 * @returns {String} The notification type, in sentence case
 */
function notificationTypeToSentence (notification) {
  const type = notification.split('_')[1];

  return `${type.slice(0, 1)}${type.slice(1).toLowerCase()}`;
}

if (typeof exports !== 'undefined') {
  exports.debug = debug;
}


js/utils.js
/*
 * This file is part of MMM-RATP (https://gitlab.com/closingin/mmm-ratp)
 * Copyright (C) 2020 Rémi Weislinger
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * debug - Displays a given text in the console if debug mode is enabled
 *
 * @param {Boolean} isDebugEnabled   Whether debug mode is enabled or not
 * @param {String}  moduleIdentifier The module identifier
 * @param {...*}    data             A variable amount of data to display in the message
 *
 * @returns {void} This function doesn't return anything
 */
function debug (isDebugEnabled, moduleIdentifier, ...data) {
  if (!isDebugEnabled) {
    return;
  }

  console.debug(`[MMM-RATP][DEBUG][${moduleIdentifier}]`, ...data);
}

/**
 * notificationTypeToLower - Retrieve a notification type and transform it to lower case
 *
 * @param {String} notification The notification
 *
 * @returns {String} The notification type, in lower case
 */
function notificationTypeToLower (notification) {
  return notification.split('_')[1].toLowerCase();
}

/**
 * notificationTypeToSentence - Retrieve a notification type and transform it to sentence case
 *
 * @param {String} notification The notification
 *
 * @returns {String} The notification type, in sentence case
 */
function notificationTypeToSentence (notification) {
  const type = notification.split('_')[1];

  return `${type.slice(0, 1)}${type.slice(1).toLowerCase()}`;
}

/**
 * isConfigurationEmpty - Check if the module config contains a valid timetables/traffic config
 *
 * @param {String} type The type to check (possible values: timetables, traffic)
 *
 * @returns {Boolean} true if the configuration is empty or invalid, false otherwise
 */
function isConfigurationEmpty(type, config) {
  if (!config[type] ||
      !Array.isArray(config[type].config) ||
      !config[type].config.length
  ) {
    return true;
  }

  return false;
}

if (typeof exports !== 'undefined') {
  exports.debug = debug;
}


node_helper.js
/*
 * This file is part of MMM-RATP (https://gitlab.com/closingin/mmm-ratp)
 * Copyright (C) 2020 Rémi Weislinger
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * @typedef FetchOptions
 * @type {Object}
 *
 * @property {Boolean} [notifyOnStart=true]  Whether to send a notification on the socket before fetching
 * @property {Boolean} [notifyOnFinish=true] Whether to send a notification on the socket after fetching
 */
const defaultFetchOptions = {
  notifyOnStart: true,
  notifyOnFinish: true
};

const NodeHelper = require('node_helper');
const RATPHelper = require('./js/ratp_helper');
const {
  debug
} = require('./js/utils');

module.exports = NodeHelper.create({
  /**
   * initializeHelper - Sets up the datastore for the given module
   *
   * @param {String} moduleIdentifier The module that requested the fetch
   * @param {Object} moduleConfig     The module's config object
   *
   * @returns {void} This function doesn't return anything
   */
  initializeHelper (moduleIdentifier, moduleConfig) {
    debug(moduleConfig.debug, moduleIdentifier, 'Initializing helper');

    this[moduleIdentifier] = {
      config: moduleConfig,
      prevData: {},
      currData: {}
    };

    this.sendSocketNotification('HELPER_INITIALIZED', { target: moduleIdentifier });
  },

  /**
   * mergeFetchOptions - Adds missing properties to the given options object
   *
   * @param {FetchOptions} options The fetch options
   *
   * @returns {void} This function doesn't return anything
   */
  mergeFetchOptions (options) {
    Object.keys(defaultFetchOptions).forEach((key) => {
      if (!(key in options)) {
        options[key] = defaultFetchOptions[key];
      }
    });
  },

  /**
   * fetchTimetables - Fetches the timetables asked by the user
   *
   * @param {String}       moduleIdentifier The module that requested the fetch
   * @param {FetchOptions} options          Fetch options
   *
   * @returns {Promise<Object[]>} A promise resolving with the fetched data
   */
  fetchTimetables (moduleIdentifier, options = defaultFetchOptions) {
    this.mergeFetchOptions(options);

    const scope = this[moduleIdentifier];
    const requests = [];

    debug(scope.config.debug, moduleIdentifier, 'Fetching timetables with options', options);

    if (options.notifyOnStart) {
      this.sendSocketNotification('UPDATING_TIMETABLES', { target: moduleIdentifier });
    }

    scope.config.timetables.config.forEach((entry) => {
      const lineType = RATPHelper.formatLineType(entry.type)

      const station = RATPHelper.apiRequest(`/stations/${lineType}/${entry.line}`)
        .then((stations) => stations.result.stations)
        .then((stations) => stations.find((s) => s.slug === entry.station));

      const timetable = RATPHelper.apiRequest(`/schedules/${lineType}/${entry.line}/${entry.station}/${entry.direction}`)
        .then((timetable) => timetable.result.schedules)
        .then((timetable) => timetable.map((nextPass) => ({
          waitingTime: RATPHelper.parseWaitingTime(nextPass.message),
          destination: nextPass.destination
        })));

      requests.push(Promise.all([
        station,
        timetable
      ]).then(([station, timetable]) => ({
        timetable,
        lineType: entry.type,
        lineName: entry.line,
        stationName: station.name,
        requestedAt: Date.now()
      })));
    });

    return Promise.all(requests).then((timetables) => {
      scope.prevData.timetables = scope.currData.timetables;
      scope.currData.timetables = timetables.map((station, idx) => {
        // NOTE: If for some unforeseen circumstances it's impossible to get a
        //       timetable, let's try to estimate it based on the last fetched
        //       one
        if (!RATPHelper.isTimetableAvailable(station.timetable)
          && scope.prevData.timetables
          && RATPHelper.isTimetableAvailable(scope.prevData.timetables[idx].timetable)
        ) {
          station.timetable = [];
          station.requestedAt = scope.prevData.timetables[idx].requestedAt;
          station.isEstimation = true;

          scope.prevData.timetables[idx].timetable.forEach((nextPass) => {
            const waitingTime = Math.round(nextPass.waitingTime - ((Date.now() - scope.prevData.timetables[idx].requestedAt) / 60000));
            station.timetable.push({ ...nextPass, waitingTime });
          });
        }

        // NOTE: Filter out values that are below zero in some cases:
        //         - when estimating next passes
        //         - when the api returns an invalid value (it can happen apparently)
        station.timetable = station.timetable.filter((nextPass) => RATPHelper.isWaitingTimeValid(nextPass.waitingTime));

        if (!station.timetable.length) {
          station.timetable.push({ waitingTime: null, destination: 'Horaires non disponibles' });
        }

        return station;
      });

      if (options.notifyOnFinish) {
        this.sendSocketNotification('DATA_TIMETABLES', {
          target: moduleIdentifier,
          payload: scope.currData.timetables
        });
      }

      return scope.currData.timetables;
    });
  },

  /**
   * fetchTraffic - Fetches the traffic asked by the user
   *
   * @param {String}       moduleIdentifier The module that requested the fetch
   * @param {FetchOptions} options          Fetch options
   *
   * @returns {Promise<Object[]>} A promise resolving with the fetched data
   */
  fetchTraffic (moduleIdentifier, options = defaultFetchOptions) {
    this.mergeFetchOptions(options);

    const scope = this[moduleIdentifier];
    const requests = [];

    debug(scope.config.debug, moduleIdentifier, 'Fetching traffic information with options', options);

    if (options.notifyOnStart) {
      this.sendSocketNotification('UPDATING_TRAFFIC', { target: moduleIdentifier });
    }

    scope.config.traffic.config.forEach((entry) => {
      requests.push(
        RATPHelper.apiRequest(`/traffic/${RATPHelper.formatLineType(entry.type)}/${entry.line}`)
          .then((traffic) => traffic.result)
          .then((traffic) => ({
            lineType: entry.type,
            lineName: entry.line,
            lineStatus: RATPHelper.parseTrafficStatus(traffic.slug),
            title: traffic.title,
            message: traffic.message
          }))
      );
    });

    return Promise.all(requests).then((traffic) => {
      scope.prevData.traffic = scope.currData.traffic;
      scope.currData.traffic = traffic;

      if (options.notifyOnFinish) {
        this.sendSocketNotification('DATA_TRAFFIC', {
          target: moduleIdentifier,
          payload: scope.currData.traffic
        });
      }

      return scope.currData.traffic;
    });
  },

  /**
   * fetchAll - Fetches all the data that the user asked for and notifies back
   *
   * @param {String} moduleIdentifier The module that requested the fetch
   *
   * @returns {Promise<Array[]>} A promise resolving with the fetched data on success
   */
  fetchAll (moduleIdentifier) {
    debug(this[moduleIdentifier].config.debug, moduleIdentifier, 'Fetching everything');

    return Promise.all([
      this.fetchTimetables(moduleIdentifier),
      this.fetchTraffic(moduleIdentifier)
    ]);
  },

  /**
   * socketNotificationReceived - See https://docs.magicmirror.builders/development/node-helper.html#socketnotificationreceived-function-notification-payload
   *
   * @param {String} notification The notification identifier
   * @param {*}      payload      The payload attached to the socket message
   *
   * @returns {void} This function doesn't return anything
   */
  socketNotificationReceived (notification, payload) {
    switch (notification) {
      case 'INITIALIZE_HELPER':
        this.initializeHelper(payload.identifier, payload.config);
        break;
      case 'FETCH_ALL':
        this.fetchAll(payload);
        break;
      case 'FETCH_TIMETABLES':
        this.fetchTimetables(payload);
        break;
      case 'FETCH_TRAFFIC':
        this.fetchTraffic(payload);
        break;
      default:
    }
  }
});


js/DOMBuilder.js
/*
 * This file is part of MMM-RATP (https://gitlab.com/closingin/mmm-ratp)
 * Copyright (C) 2020 Rémi Weislinger
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * DOMBuilder - Dedicated methods to build the DOM of this MagicMirror module
 */
class DOMBuilder {
  /**
   * @static createTimetables - Create a timetables block element
   *
   * @param {Object}    _         An object containing the timetables information
   * @param {Object[]}  _.entries   The current timetables
   * @param {Object}    _.status    Additional information about the timetables
   * @param {Object}   config     The module configuration
   *
   * @returns {HTMLElement} The newly created timetables block
   */
  static createTimetables ({ entries, status }, config) {
    const section = DOMBuilder.createSection('timetables', status, config);

    entries.forEach((station) => {
      const card = DOMBuilder.createCard(
        DOMBuilder.createLine(station.lineType, station.lineName),
        {
          title: station.stationName,
          subtitle: `&gt; ${station.timetable[0].destination}`
        },
        DOMBuilder.createTimetable(station.timetable.slice(0, config.timetables.nextPassesAmount), station.estimation)
      );

      section.appendChild(card);
    });

    return section;
  }

  /**
   * @static createTraffic - Create a traffic block element
   *
   * @param {Object}    _         An object containing the traffic information
   * @param {Object[]}  _.entries   The current traffic
   * @param {Object}    _.status    Additional information about the traffic
   * @param {Object}   config     The module configuration
   *
   * @returns {HTMLElement} The newly created traffic block
   */
  static createTraffic ({ entries, status }, config) {
    const section = DOMBuilder.createSection('traffic', status, config);

    entries.forEach((info) => {
      if (info.lineStatus === 'normal' && config.traffic.hideWhenNormal) {
        return;
      }

      const card = DOMBuilder.createCard(
        DOMBuilder.createLine(info.lineType, info.lineName, info.lineStatus),
        {
          title: info.title,
          content: info.message
        }
      );

      section.appendChild(card);
    });

    if (section.children.length === 1) {
      section.appendChild(DOMBuilder.createCard(null, { content: 'Trafic normal sur l\'ensemble des lignes spécifiées' }));
    }

    return section;
  }

  /**
   * @static createSection - Create a section element
   *
   * @param {String} type   The type of content that will be in the section
   * @param {Object} status Additional information about the content
   * @param {Object} config The module configuration
   *
   * @returns {HTMLElement} The newly created section
   */
  static createSection (type, status, config) {
    const element = document.createElement('div');

    element.className = `MMM-RATP__section MMM-RATP__${type}`;
    element.innerHTML = `<h3 class="MMM-RATP__section__title">${config[type].title}</h3>`;

    if (status.isUpdating && config.showUpdateAnimation) {
      const spinner = document.createElement('span');

      spinner.className = 'MMM-RATP__spinner';
      element.querySelector('.MMM-RATP__section__title').appendChild(spinner);
    }

    return element;
  }

  /**
   * @static createCard - Create a card element
   *
   * @param {HTMLElement} [prepend=null] The element to prepend to the card (a line logo)
   * @param {Object}      [text=null]    The text to put in the middle of the card
   * @param {String}      text.title       The title of the card
   * @param {String}      text.subtitle    The subtitle of the card
   * @param {String}      text.content     The content of the card
   * @param {HTMLElement} [append=null]  The element to append to the card (a timetable for example)
   *
   * @returns {HTMLElement} The newly created card
   */
  static createCard (prepend = null, text = null, append = null) {
    const element = document.createElement('div');
    element.className = 'MMM-RATP__card';

    if (prepend) {
      prepend.classList.add('MMM-RATP__card__prepend');
      element.appendChild(prepend);
    }

    if (text) {
      const textElement = document.createElement('div');

      textElement.className = 'MMM-RATP__card__text';
      textElement.innerHTML = `
        ${(text.title)    ? `<h4 class="MMM-RATP__card__text__title">${text.title}</h4>` : ''}
        ${(text.subtitle) ? `<h5 class="MMM-RATP__card__text__subtitle">${text.subtitle}</h5>` : ''}
        ${(text.content)  ? `<p class="MMM-RATP__card__text__content">${text.content}</p>` : ''}
      `;

      element.appendChild(textElement);
    }

    if (append) {
      append.classList.add('MMM-RATP__card__append');
      element.appendChild(append);
    }

    return element;
  }

  /**
   * @static createLine - Create a line element
   *
   * @param {String} type          The type of the line (metro, tramway, ...)
   * @param {String} name          The name of the line (1, A, 3b, ...)
   * @param {String} [status=null] The current traffic status on the line
   *
   * @returns {HTMLElement} The newly created line
   */
  static createLine (type, name, status = null) {
    const element = document.createElement('div');

    element.className = `MMM-RATP__line MMM-RATP__line--${type}`;
    element.innerHTML = name;
    element.setAttribute('data-line', name.toLowerCase());

    if (status) {
      element.classList.add('MMM-RATP__line--badge');
      element.classList.add(`MMM-RATP__line--badge-${status}`);
    }

    return element;
  }

  /**
   * @static createTimetable - Create a timetable element
   *
   * @param {Object[]} timetable  The timetable data
   * @param {Boolean}  estimation Whether the times in the timetable are an estimation or not
   *
   * @returns {HTMLElement} The newly created timetable
   */
  static createTimetable (timetable, estimation) {
    const element = document.createElement('div');
    element.className = 'MMM-RATP__timer-container';

    timetable.forEach((timetableEntry) => {
      const timer = DOMBuilder.createTimer(
        timetableEntry.waitingTime,
        estimation
      );

      element.appendChild(timer);
    });

    return element;
  }

  /**
   * @static createTimer - Create a timer element
   *
   * @param {String}  time       The timer's content
   * @param {Boolean} estimation Whether the given time is an estimation or not
   *
   * @returns {HTMLElement} The newly created timer
   */
  static createTimer (time, estimation) {
    const element = document.createElement('div');

    element.className = 'MMM-RATP__timer';
    element.innerHTML = time || '?';

    if (time === '0') element.classList.add('MMM-RATP__timer--incoming');
    if (estimation)   element.innerHTML += '*';

    return element;
  }
}


MMM-RATP.js
/*
 * This file is part of MMM-RATP (https://gitlab.com/closingin/mmm-ratp)
 * Copyright (C) 2020 Rémi Weislinger
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

Module.register('MMM-RATP', {
  /**
   * requiresVersion - Minimum MagicMirror² version required to use this module
   *
   * @type {String}
   */
  requiresVersion: '2.1.0',

  /**
   * defaults - The default data that this module will use
   *
   * @type {Object}
   */
  defaults: {
    theme: 'mirror',
    debug: false,
    showUpdateAnimation: true,

    timetables: {
      title: 'Prochains passages',
      updateInterval: 1 * 60 * 1000, // Every minute, in milliseconds
      config: [],
      nextPassesAmount: 2
    },

    traffic: {
      title: 'Infos trafic',
      updateInterval: 10 * 60 * 1000, // Every 10 minutes, in milliseconds
      config: [],
      hideWhenNormal: false
    }
  },

  /**
   * setDefaults - Sets the default values for timetables and traffic if the user didn't set them
   *
   * @returns {void} This function doesn't return anything
   */
  setDefaults () {
    Object.keys(this.defaults.timetables).forEach((option) => {
      if (!(option in this.config.timetables)) {
        this.config.timetables[option] = this.defaults.timetables[option];
      }
    });

    Object.keys(this.defaults.traffic).forEach((option) => {
      if (!(option in this.config.traffic)) {
        this.config.traffic[option] = this.defaults.traffic[option];
      }
    });
  },

  /**
   * initializeModule - Registers the functions to call on refresh and registers the module on the helper
   *
   * @returns {void} This function doesn't return anything
   */
  initializeModule () {
    debug(this.config.debug, this.identifier, `Initializing module`);

    if (!isConfigurationEmpty('timetables', this.config)) {
      setInterval(() => {
        this.sendSocketNotification('FETCH_TIMETABLES', this.identifier);
      }, this.config.timetables.updateInterval);
    }

    if (!isConfigurationEmpty('traffic', this.config)) {
      setInterval(() => {
        this.sendSocketNotification('FETCH_TRAFFIC', this.identifier);
      }, this.config.traffic.updateInterval);
    }

    this.sendSocketNotification('INITIALIZE_HELPER', {
      identifier: this.identifier,
      config: this.config
    });
  },

  /**
   * start - See https://docs.magicmirror.builders/development/core-module-file.html#start
   *
   * @returns {void} This function doesn't return anything
   */
  start () {
    this.apiData = {
      timetables: { entries: [], status: {} },
      traffic:    { entries: [], status: {} }
    };

    this.setDefaults();
    this.initializeModule();
  },

  /**
   * getScripts - See https://docs.magicmirror.builders/development/core-module-file.html#getscripts
   *
   * @returns {String[]} The scripts to load
   */
  getScripts () {
    return [
      this.file('js/utils.js'),
      this.file('js/dom_builder.js')
    ];
  },

  /**
   * getStyles - See https://docs.magicmirror.builders/development/core-module-file.html#getstyles
   *
   * @returns {String[]} The styles to load
   */
  getStyles () {
    // NOTE: Add your custom theme to this array in order to be able to load it
    const themes = [
      'mirror',
      'dashboard'
    ];

    if (themes.indexOf(this.config.theme) === -1) {
      Log.error(`${this.name}: invalid theme, did you forget to add it to MMM-RATP::getStyles() ?`);
      return;
    }

    return [
      this.file(`${this.name}.css`),
      this.file(`css/${this.name}_positioning.css`),
      this.file(`css/${this.name}_components.css`),
      this.file(`css/themes/${this.name}_theme_${this.config.theme}.css`)
    ];
  },

  /**
   * getDom - See https://docs.magicmirror.builders/development/core-module-file.html#getdom
   *
   * @returns {DocumentFragment} The fragment that will be added to .module-content
   */
  getDom () {
    debug(this.config.debug, this.identifier, `Rebuilding the DOM`);

    const fragment = document.createDocumentFragment();

    if (!isConfigurationEmpty('timetables', this.config)) {
      fragment.append(DOMBuilder.createTimetables(this.apiData.timetables, this.config));
    }

    if (!isConfigurationEmpty('traffic', this.config)) {
      fragment.append(DOMBuilder.createTraffic(this.apiData.traffic, this.config));
    }

    return fragment;
  },

  /**
   * socketNotificationReceived - See https://docs.magicmirror.builders/development/core-module-file.html#socketnotificationreceived-function-notification-payload
   *
   * @param {String} notification The notification identifier
   * @param {Object} _            The payload attached to the socket message
   * @param {String} _.target       The target of the notification (the module identifier)
   * @param {*}      _.payload      The actual payload for the notification
   *
   * @returns {void} This function doesn't return anything
   */
  socketNotificationReceived (notification, { target, payload }) {
    // NOTE: Socket messages are broadcasted to every instantiated modules, so
    //       we need to check whether we are the recipient of a message or not
    if (this.identifier !== target) {
      return;
    }

    debug(this.config.debug, this.identifier, `Received socket message:\n  - identifier: ${notification}\n  - payload: ${payload}`);

    switch (notification) {
      case 'HELPER_INITIALIZED':
        this.sendSocketNotification('FETCH_ALL', this.identifier);
        return;
      case 'DATA_TIMETABLES':
      case 'DATA_TRAFFIC':
        const type = notificationTypeToLower(notification);

        this.apiData[type].status.isUpdating = false;
        this.apiData[type].status.lastUpdate = new Date();
        this.apiData[type].entries = payload;
        break;
      case 'UPDATING_TIMETABLES':
      case 'UPDATING_TRAFFIC':
        this.apiData[notificationTypeToLower(notification)].status.isUpdating = true;
        break;
    }

    this.updateDom();
  }
});
